docs/docs/icicle/primitives/merkle.md

```diff
@@ -1,3 +1,252 @@
+2. **Size of a leaf element** (in bytes): This defines the **granularity** of the data used for opening proofs.
+
+---
+
+## Tree Structure and Configuration
+
+### Structure Definition
+
+The **root node** is assumed to be a single node. The **height of the tree** is determined by the **number of layers**.  
+Each layer's **arity** is calculated as:
+
+$$
+{arity}_i = \frac{layers[i].inputSize}{layer[i-1].outputSize}
+$$
+
+For **layer 0**:
+
+$$
+{arity}_0 = \frac{layers[0].inputSize}{leafSize}
+$$
+
+---
+
+### Defining a Merkle Tree
+
+```cpp
+// icicle/merkle/merkle_tree.h
+static MerkleTree create(
+    const std::vector<Hash>& layer_hashes, 
+    uint64_t leaf_element_size, 
+    uint64_t output_store_min_layer = 0);
+```
+
+---
+
+### Building the Tree
+
+```cpp
+// icicle/merkle/merkle_tree.h
+inline eIcicleError build(
+    const std::byte* leaves, 
+    uint64_t leaves_size, 
+    const MerkleTreeConfig& config);
+
+template <typename T>
+inline eIcicleError build(
+    const T* leaves, 
+    uint64_t nof_leaves, 
+    const MerkleTreeConfig& config);
+```
+
+---
+
+## Tree Examples
+
+### Example A: Binary Tree
+
+A binary tree with **5 layers**, using **Keccak-256**:
+
+![Merkle Tree Diagram](./merkle_diagrams/diagram1.png)
+
+```cpp
+const uint64_t leaf_size = 1024;
+// Allocate a dummy input. It can be any type as long as the total size matches.
+const uint32_t max_input_size = leaf_size * 16;
+auto input = std::make_unique<uint64_t[]>(max_input_size / sizeof(uint64_t));
+
+// Define hasher
+auto layer0_hasher = Keccak256::create(leaf_size); // hash 1KB -> 32B
+auto next_layer_hasher = Keccak256::create(2 * layer0_hasher.output_size()); // hash every 64B to 32B
+
+// Construct the tree using the layer hashes and leaf-size
+std::vector<Hash> hashes = {layer0_hasher, next_layer_hasher, next_layer_hasher, next_layer_hasher, next_layer_hasher};
+auto merkle_tree = MerkleTree::create(hashes, leaf_size);
+
+// compute the tree
+merkle_tree.build(input.get(), max_input_size / sizeof(uint64_t), default_merkle_tree_config());
+```
+
+---
+
+### Example B: Tree with Arity 4
+
+This example uses **Blake2s** in the upper layer:
+
+![Merkle Tree Diagram](./merkle_diagrams/diagram2.png)
+
+```cpp
+#include "icicle/merkle/merkle_tree.h"
+
+const uint64_t leaf_size = 1024;
+const uint32_t max_input_size = leaf_size * 16;
+auto input = std::make_unique<uint64_t[]>(max_input_size / sizeof(uint64_t));
+
+// note here we use Blake2S for the upper layer
+auto layer0_hasher = Keccak256::create(leaf_size);
+auto next_layer_hasher = Blake2s::create(4 * layer0_hasher.output_size());
+
+std::vector<Hash> hashes = {layer0_hasher, next_layer_hasher};
+auto merkle_tree = MerkleTree::create(hashes, leaf_size);
+
+merkle_tree.build(input.get(), max_input_size / sizeof(uint64_t), default_merkle_tree_config());
+```
+
+:::note
+Any combination of hashes is valid including **Poseidon** that computes on field elements.
```

Comment:
Are there any limitations on Poseidon in the tree (Odd bit width for example messing with the layer width)?

Reply:
I think that the field size must divide the output size of previous layer. But this is true for all other hashes too.
About converting random data to field elements it's the responsibility of the hasher function, not the tree.